pkg/sql/compile/ddl.go

```diff
@@ -108,25 +121,131 @@ func (s *Scope) CreateTable(c *Compile) error {
 	return colexec.CreateAutoIncrCol(c.e, c.ctx, dbSource, c.proc, tableCols, dbName, tblName)
 }
 
+func (s *Scope) CreateTempTable(c *Compile) error {
+	qry := s.Plan.GetDdl().GetCreateTable()
+	// convert the plan's cols to the execution's cols
+	planCols := qry.GetTableDef().GetCols()
+	tableCols := planCols
+	exeCols := planColsToExeCols(planCols)
+
+	// convert the plan's defs to the execution's defs
+	planDefs := qry.GetTableDef().GetDefs()
+	exeDefs, err := planDefsToExeDefs(planDefs)
+	if err != nil {
+		return err
+	}
+
+	// Temporary table names and persistent table names are not allowed to be duplicated
+	// So before create temporary table, need to check if it exists a table has same name
+	dbName := c.db
+	if qry.GetDatabase() != "" {
+		dbName = qry.GetDatabase()
+	}
+
+	// check in EntireEngine.TempEngine
+	tmpDBSource, err := c.e.Database(c.ctx, engine.TEMPORARY_DBNAME, c.proc.TxnOperator)
+	if err != nil {
+		return err
+	}
+	tblName := qry.GetTableDef().GetName()
+	if _, err := tmpDBSource.Relation(c.ctx, dbName+"-"+tblName); err == nil {
+		if qry.GetIfNotExists() {
+			return nil
+		}
+		return moerr.NewTableAlreadyExists(fmt.Sprintf("temporary '%s'", tblName))
+	}
+
+	// check in EntireEngine.Engine
+	dbSource, err := c.e.Database(c.ctx, dbName, c.proc.TxnOperator)
+	if err != nil {
+		return err
+	}
+	if _, err := dbSource.Relation(c.ctx, tblName); err == nil {
+		if qry.GetIfNotExists() {
+			return nil
+		}
+		return moerr.NewTableAlreadyExists(tblName)
+	}
+
+	// create temporary table
+	if err := tmpDBSource.Create(c.ctx, dbName+"-"+tblName, append(exeCols, exeDefs...)); err != nil {
+		return err
+	}
+
+	// build index table
+	for _, def := range qry.IndexTables {
+		planCols = def.GetCols()
+		exeCols = planColsToExeCols(planCols)
+		planDefs = def.GetDefs()
+		exeDefs, err = planDefsToExeDefs(planDefs)
+		if err != nil {
+			return err
+		}
+		if _, err := tmpDBSource.Relation(c.ctx, def.Name); err == nil {
+			return moerr.NewTableAlreadyExists(def.Name)
+		}
+		if err := tmpDBSource.Create(c.ctx, dbName+"-"+def.Name, append(exeCols, exeDefs...)); err != nil {
```

Comment:
this is crazy.    at least it has to be one function in one place.

Reply:
So is this a solution that you think is passable? if so I will fix it tomorrow.